kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1398 @@
+These applications usually include tens to hundreds of microservices with complex interdependencies/chains.
+Distance from servers with chained microservices deployed is usually the primary culprit.
+The best strategy is to reduce the latency between chained microservices in the same application,
+according to the prior work about [Service Function Chaining](https://www.sciencedirect.com/science/article/pii/S1084804516301989) (SFC).
+
+Besides, bandwidth plays an essential role for those applications with high volumes of data transfers among microservices.
+For example, multiple replicas in a database application may require frequent copies to ensure data consistency.
+[Spark jobs](https://spark.apache.org/) may have frequent data transfers among map and reduce nodes.
+Insufficient network capacity in nodes would lead to increasing delay or packet drops,
+which will degrade the quality of service for applications.
+
+We propose a series of **Network-Aware Scheduling Plugins** for Kubernetes that focus on delivering low latency to end-users
+and ensuring bandwidth reservations in pod scheduling.
+
+This work significantly extends the previous work open-sourced [here](https://github.com/jpedro1992/sfc-controller) 
+that implements a latency-aware scheduler extender based on the [scheduler extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md) design.
+
+## Goals
+
+- Provide a network-aware framework to extend scheduling features of Kubernetes by considering latency and bandwidth.
+- Consider different pods as an Application:    
+    - The creation of an **Application Group (AppGroup) CRD**.
+- Define network weights between regions (`topology.kubernetes.io/region`) and zones (`topology.kubernetes.io/zone`) in the cluster:
+    - The creation of a **Network Topology (NetworkTopology) CRD**.
+- The advertising of the nodes (physical) bandwidth capacity as [extended resources](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/): 
+    - Bandwidth requests and limitations allow filtering overloaded nodes (bandwidth) considered for scheduling.
+    - Consider bandwidth requests for scoring plugins (e.g., `MostRequested`, `BalancedAllocation`). 
+- Establish a specific order to allocate Pods based on their AppGroup CRD:
+    - Implementation of a **QueueSort** plugin based on [Topology Sorting](https://en.wikipedia.org/wiki/Topological_sorting#:~:text=In%20computer%20science%2C%20a%20topological,before%20v%20in%20the%20ordering).
+- Filter out nodes based on microservice dependencies among pods belonging to the same application:
+    - Implementation of a **Filter** plugin based on the AppGroup CRD.
+- Near-optimal scheduling decisions based on network costs:
+    - Implementation of a **Score** plugin that scores higher nodes with lower network costs.    
+
+## Non-Goals
+
+- Descheduling due to unexpected outcomes is not addressed in this proposal.
```

Comment:
Could you elaborate a little more on this:
1. What unexpected outcomes did you mean
2. Do you have some observations or tests

Reply:
1) For instance, we plan to perform experiments to evaluate if our plugins increase significantly resource fragmentation. It is expected that further nodes would be used for pod deployments since we try to reduce the latency. So, we plan to measure the impact concerning fragmentation since it can lead to descheduling. As future improvements, we plan to target this challenge as well. 

2) Not yet, but we plan to perform experiments and assess the impact as mentioned above.